lisa-kernel/src/main/scala/lisa/kernel/fol/Substitutions.scala

```diff
@@ -195,4 +195,44 @@ trait Substitutions extends FormulaDefinitions {
     }
   }
 
+  /**
+   * Instantiate a schematic connector symbol in a formula, using higher-order instantiation.
+   *
+   * @param phi The base formula
+   * @param m   The map from schematic function symbols to lambda expressions Formula(s) -> Formula [[LambdaFormulaFormula]].
+   * @return phi[m]
+   */
+  def instantiateSchemas(
+      phi: Formula,
+      mCon: Map[SchematicConnectorLabel, LambdaFormulaFormula],
+      mPred: Map[SchematicVarOrPredLabel, LambdaTermFormula],
+      mTerm: Map[SchematicTermLabel, LambdaTermTerm]
+  ): Formula = {
+    require(mCon.forall { case (symbol, LambdaFormulaFormula(arguments, body)) => arguments.length == symbol.arity })
+    require(mPred.forall { case (symbol, LambdaTermFormula(arguments, body)) => arguments.length == symbol.arity })
+    require(mTerm.forall { case (symbol, LambdaTermTerm(arguments, body)) => arguments.length == symbol.arity })
+    phi match {
+      case PredicateFormula(label, args) =>
+        label match {
+          case label: SchematicVarOrPredLabel if mPred.contains(label) => mPred(label)(args)
```

Comment:
Doing some more tests, it seems this line causes simultaneous substitution to fail

```scala
val phi = P(f(x))

val mPred = Map(P -> LambdaTermFormula(Seq(x), Q(g(x)))
val mFun = Map(f -> LambdaTermTerm(Seq(x), f(f(x))))

instantiateSchemas(phi, Map.empty, mPred, mFun)
// Output   : Q(g(f(x)))
// Expected : Q(g(f(f(x))))

Reply:
This works now